### PR TITLE
Deprecate deepspeed

### PR DESCRIPTION
**Pull request by dakinggg**
# What does this PR do?
This PR deprecates the usage of DeepSpeed in Composer. We have not updated DeepSpeed in a long time and do not actively maintain the integration. Composer is tightly coupled to PyTorch FSDP which provides similar functionality, and we are developing all new functionality on top of FSDP and PyTorch native, not DeepSpeed, for parallelism. Before removal we will provide a migration guide for anyone still using DeepSpeed. In the meantime, users are welcome to open a GitHub issue if they have any trouble.

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -1106,6 +1106,14 @@ def __init__(
         # compile config for PyTorch 2.0 or higher
         compile_config: Optional[dict[str, Any]] = None,
     ):
+        if deepspeed_config is not None:
+            raise VersionedDeprecationWarning(
+                'The use of DeepSpeed in Composer is deprecated. Composer is tightly integrated with PyTorch FSDP ' +
+                'which provides similar functionality. Please use the `parallelism_config` parameter instead. Please open '
+                + 'a GitHub issue if you need help migrating from DeepSpeed to FSDP.',
+                remove_version='0.28.0',
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1114]:
```suggestion
                remove_version='1.0.0',
```
We can tighten later

Comment (dakinggg) [reply composer/trainer/trainer.py:1114]: We will plan on removing support for training new models using deepspeed in the version list, but will keep support for loading existing deepspeed models until version 1.0.

Comment (dakinggg) [reply composer/trainer/trainer.py:1114]: Updated message to be more specific to what will be removed in 0.28

**Review by mvpatel2000 — APPROVED** [review]
LGTM

**Review by abhi-mosaic — APPROVED** [review]
🫡

**Review by bandish-shah — APPROVED** [review]
FSDP is all you need

**Review by jbloxham — APPROVED** [review]
Two and a half years later and I still am confused by DeepSpeed checkpointing

**Review by snarayan21 — COMMENTED** [review]
seems that all the deepspeed related tests are failing

**Pull request merged**